flang/include/flang/Optimizer/Dialect/FIRType.h

```diff
@@ -97,6 +97,10 @@ bool isa_aggregate(mlir::Type t);
 /// not a memory reference type, then returns a null `Type`.
 mlir::Type dyn_cast_ptrEleTy(mlir::Type t);
 
+/// Extract the `Type` pointed to from a FIR memory reference or box type. If
+/// `t` is not a memory reference or box type, then returns a null `Type`.
+mlir::Type dyn_cast_ptrOrBoxEleTy(mlir::Type t);
```

Comment:
I've wondered if `dyn_cast_ptrEleTy` shouldn't already work this way with box type.

Reply:
There are about 30 usages of `dyn_cast_ptrEleTy` in lowering/codegen, and most them are in places where fir.box would not make sense/be allowed, so I think a new one is safer.